### ISSUE TITLE
There is no support for multiple response types

### ISSUE THREAD

**Issue by JKostov** [IBM/openapi-to-graphql#378]
**Describe the bug**
When using multiple response types it will use the first one or application/json if defined.

**To Reproduce**

Use the following schema:

```yaml
openapi: 3.0.3
info:
  version: v1
  title: Multiple response type example
paths:
  "/test":
    get:
      summary: Get value
      responses:
        "200":
          description: Returns value
          content:
            text/plain:
              example: example
            application/json:
              schema:
                type: object
```

**Expected behavior**
It should create union type from the specified response types

**Comment by Alan-Cha** [comment]
So far, this has been a deliberate choice as GraphQL normally returns JSON so we prioritize application/json.

Besides text/plain, are there other media types that you think we should be supporting?

Assuming that text/plain should be translated into a string, then in this case, a union would actually not work because according to the spec, a GraphQL union cannot be formed with scalars. The best we can do is wrap it all in a GraphQLJSON, which isn't a great alternative because we lose a lot of the typing benefits of GraphQL.

**Comment by vasicvuk** [comment]
Wrapping text/plain response in JSON would be good enough i think. OpenAPI can also work with XML (conversion to JSON) is also option.

One more scenario are response that are files. Maybe we can return files as object having file property which is a base64 string of file.

Another possibility is for object to have multiple properties such as (content, contentType, contentSize, name)

**Comment by Alan-Cha** [comment]
@vasicvuk Thanks for your input.

> Wrapping text/plain response in JSON would be good enough i think.

Just to clarify, I think text/plain should be string but if we want to support both text/plain and application/json from the same operation, then one way we could do it is to wrap it all in JSON. Is that what you mean?

> Maybe conversion to JSON is also option for XML.

Yes, this could be a future work.

> Another possibility is for object to have multiple properties such as (content, contentType, contentSize, name)

This is an interesting take but I think one of the advantages of GraphQL is that it abstracts away what's going on in the backend. I don't think we necessarily want all of this in OtG. What do you think?

**Comment by vasicvuk** [comment]
> @vasicvuk Thanks for your input.
> 
> > Wrapping text/plain response in JSON would be good enough i think.
>
> Just to clarify, I think text/plain should be string but if we want to support both text/plain and application/json from the same operation, then one way we could do it is to wrap it all in JSON. Is that what you mean?
> 

Sure that is what i suggested. Only in case of combination between plain text and json then return plain text wrapped in JSON.

> > Maybe conversion to JSON is also option for XML.
> 
> Yes, this could be a future work.
> 

Maybe i should open another issue (feature request) for this?

> > Another possibility is for object to have multiple properties such as (content, contentType, contentSize, name)
> 
> This is an interesting take but I think one of the advantages of GraphQL is that it abstracts away what's going on in the backend. I don't think we necessarily want to expose all of this in OtG. What do you think?

Maybe as an option we can have possibility to register some custom converters for different formats (different content types). This will allow users to create custom rules not touching the source code of OtG.

**Comment by Alan-Cha** [comment]
@vasicvuk Do you see this as a very likely use case? To support different media types for the same operation when we can prioritize JSON? I am also uncertain of how appropriate it is to send media types outside of plain text and JSON such as files. Perhaps for those kinds of APIs, GraphQL isn't well-suited.

In addition, how would a user select between different media types? Perhaps we could add a new argument for operations with multiple media types and that would affect the `Accept` header of the API call.

> Maybe i should open another issue (feature request) for this?

Yes, please. In addition to XML, are there other formats that we should support?

> Maybe as an option we can have possibility to register some custom converters for different formats (different content types). This will allow users to create custom rules not touching the source code of OtG.

I don't completely follow. Are you suggesting letting users define logic to create their own object types and fields? I don't think this is likely. This can get quite complicated with nested types and moreover, resolvers are created separately and need to match the object types, so there is complex logic there that users will need to define as well.

In any case, I think there are more fundamental questions to answer such as the ones I mentioned in the beginning before we can even consider this.

**Comment by vasicvuk** [comment]
> @vasicvuk Do you see this as a very likely use case? To support different media types for the same operation when we can prioritize JSON? I am also uncertain of how appropriate it is to send media types outside of plain text and JSON such as files. Perhaps for those kinds of APIs, GraphQL isn't well-suited.
> 
> In addition, how would a user select between different media types? Perhaps we could add a new argument for operations with multiple media types and that would affect the `Accept` header of the API call.
> 

I think the issue is when user does not send the Accept header to API but backend API returns different media type based on the logic that is server side. Same endpoint can return different media types based on some logic. This media type can be json, xml, plain text, csv.

Accept header is not always processed by implementation of API so this would be hard requirement.

> > Maybe i should open another issue (feature request) for this?
> 
> Yes, please. In addition to XML, are there other formats that we should support?
> 
> > Maybe as an option we can have possibility to register some custom converters for different formats (different content types). This will allow users to create custom rules not touching the source code of OtG.
> 
> I don't completely follow. Are you suggesting letting users define logic to create their own object types and fields? I don't think this is likely. This can get quite complicated with nested types and moreover, resolvers are created separately and need to match the object types, so there is complex logic there that users will need to define as well.
> 
> In any case, I think there are more fundamental questions to answer such as the ones I mentioned in the beginning before we can even consider this.

**Comment by Alan-Cha** [comment]
Yes, I see. Currently, OtG will just ignore the response if it is not of the same media type that was requested.

> Accept header is not always processed by implementation of API so this would be hard requirement.

How common is this though? And should OtG handle these kinds of unpredictable APIs? We've spent a lot of debating over similar topics in the past. whether the OAS is poorly written or the API works unpredictably, we have to draw a line somewhere. We want to support the majority of APIs and if a few do not work well without multiple compromises, then that is fine. In a more practical sense, there may be more important features to work on then to ensure we can cover all APIs.

I'm not necessarily making a judgement on this issue. I am still looking to be convinced about unpredictable APIs and also support other media types like files.

**Comment by vasicvuk** [comment]
Hi @Alan-Cha,

I can tell you that this scenario is very common. These APIs are not unpredictable. They are very predictable and defined using OpenAPI standard to return multiple type of Content Types. 

I see two right paths for **OtG** here:
- Add native support for multiple content types
- Add possibility for OtG users to extend OtG with custom Content Type mappers 

Other path will lead to creating forks whenever something from OpenAPI is not supported in OtG
